lib/pure/collections/deques.nim

```diff
@@ -52,46 +51,69 @@
 ## * `channels module <channels.html>`_ for inter-thread communication
 
 
-import math
+import math, hashes
 
 type
   Deque*[T] = object
     ## A double-ended queue backed with a ringed seq buffer.
     ##
-    ## To initialize an empty deque use `initDeque proc <#initDeque,int>`_.
+    ## To initialize an empty deque with a given capacity use
+    ## `initDeque proc <#initDeque,int>`_.
     data: seq[T]
     head, tail, count, mask: int
 
 const
-  defaultInitialSize* = 4
+  dequeDefaultInitialCapacity* {.intdefine.} = 4
+
+template initImpl(result: typed, initialCapacity: int) =
+  assert isPowerOfTwo(initialCapacity)
+  result.mask = initialCapacity-1
+  newSeq(result.data, initialCapacity)
 
-template initImpl(result: typed, initialSize: int) =
-  assert isPowerOfTwo(initialSize)
-  result.mask = initialSize-1
-  newSeq(result.data, initialSize)
-  
 template checkIfInitialized(deq: typed) =
-  when compiles(defaultInitialSize):
+  when declared(dequeDefaultInitialCapacity):
     if deq.mask == 0:
-      initImpl(deq, defaultInitialSize)
+      initImpl(deq, dequeDefaultInitialCapacity)
 
-proc initDeque*[T](initialSize: int = 4): Deque[T] =
-  ## Create a new empty deque.
+proc initDeque*[T](initialCapacity: int = dequeDefaultInitialCapacity): Deque[T] =
+  ## Create a new empty deque with a given capacity. An implicitly defined
+  ## deque will have a capacity of 0 and be grown to fit elements on the first
+  ## data insertion.
   ##
-  ## Optionally, the initial capacity can be reserved via `initialSize`
-  ## as a performance optimization.
-  ## The length of a newly created deque will still be 0.
+  ## Optionally, the initial capacity can be reserved via `initialCapacity`
+  ## as a performance optimization. The length of a newly created deque will
+  ## still be 0.
   ##
-  ## ``initialSize`` must be a power of two (default: 4).
+  ## ``initialCapacity`` must be a power of two (default: 4).
   ## If you need to accept runtime values for this you could use the
   ## `nextPowerOfTwo proc<math.html#nextPowerOfTwo,int>`_ from the
   ## `math module<math.html>`_.
-  result.initImpl(initialSize)
+  result.initImpl(initialCapacity)
 
 proc len*[T](deq: Deque[T]): int {.inline.} =
-  ## Return the number of elements of `deq`.
+  ## Return the number of elements in the `deq`.
   result = deq.count
 
+template high*[T](deq: Deque[T]): int =
+  deq.len - 1
+
+proc toDeque*[T](x: seq[T]): Deque[T] =
+  ## Converts a sequence into a Deque
+  runnableExamples:
+    var x = @[10, 20, 30].toDeque
+    assert x.len == 3
+    assert x[0] == 10
+    x.addFirst 0
+    x.addLast 40
+    assert $x == "[0, 10, 20, 30, 40]"
+  result.head = 0
+  result.count = x.len
+  result.tail = x.len
+  result.data = x
```

Comment:
not efficient as this can reallocate un-necessarily
instead use smthg like this:
```nim
if x.len.isPowerOfTwo:
  result.data = x
else:
  newSeqOfCap(result.data, x.len.nextPowerOfTwo)
  result.data.add x
```

Reply:
How would that have less overhead? You're allocating a completely new memory region with `newSeqOfCap` and then copying all the memory in there. With my solution the `setLen` call could trigger a `realloc` which could allow the OS to just assign more memory in the same region and not copy anything. Or am I missing something here?

Reply:
in your code:
* `result.data = x` does a copy, it only requests for a memory region with x.len elements, so could be (if memory is fragmented) allocating it in a region with less than x.len.nextPowerOfTwo elements.
* then if `not x.len.isPowerOfTwo` (arguably, the common case), `result.data.setLen` will call realloc, which may have to allocate to a new region (as explained above)

in my code: I only allocate exactly what's needed ahead of time, and there is 0 overhead, the above situation can't happen, since I'm directly allocating exactly what's needed in the end. No spurious copies either, since `newSeqOfCap` IIUC shouldn't do any zero-ing out of memory. The only zeroing out of memory is in `result.data.setLen n` but it zeros out exactly what's needed, nothing more.

Reply:
Oops, forgot that `result.data = x` would create a copy. Is this the case even when passed a literal? TBH I wasn't overly concerned with performance for this procedure when I wrote it as I considered the most common use case to be things like `let x = @[1, 2, 3].toDeque` when testing simple things.

Reply:
nah, passing a seq that's not a litteral is a perfectly sensible use case; in fact, I now realize that this should be openArray, not seq! Since this is a new API, we should do the right thing here and use `openArray`.

just replace
```
result.data = x
```
by
```
result.data.add x
```
in my snippet